src/InstrumentationEngine/ProfilerManager.cpp

```diff
@@ -196,50 +196,106 @@ HRESULT CProfilerManager::AddRawProfilerHook(
 
     unique_ptr<CProfilerCallbackHolder> profilerCallbackHolder(new CProfilerCallbackHolder);
 
-    CComPtr<ICorProfilerCallback> pCorProfilerCallback;
-    IfFailRet(pUnkProfilerCallback->QueryInterface(__uuidof(ICorProfilerCallback), (LPVOID*)&pCorProfilerCallback));
-    profilerCallbackHolder->m_CorProfilerCallback = pCorProfilerCallback;
+    // Follow the implementation set by the CLR at
+    // https://github.com/dotnet/runtime/blob/cf6b06b1d36d545e37b00bf1a6311b7fff33ff4e/src/coreclr/src/vm/eetoprofinterfaceimpl.cpp#L613
+    // where due to interface inheritance, higher numbered ICorProfilerCallback## are static-casted to lower numbered ICorProfilerCallback##
 
-    CComPtr<ICorProfilerCallback2> pCorProfilerCallback2;
-    hr = pUnkProfilerCallback->QueryInterface(__uuidof(ICorProfilerCallback2), (LPVOID*)&pCorProfilerCallback2);
+    // ICorProfilerCallback7
+    CComPtr<ICorProfilerCallback7> pCorProfilerCallback7;
+    hr = pUnkProfilerCallback->QueryInterface(__uuidof(ICorProfilerCallback7), (LPVOID*)&pCorProfilerCallback7);
     if (SUCCEEDED(hr))
     {
-        profilerCallbackHolder->m_CorProfilerCallback2 = pCorProfilerCallback2;
+        profilerCallbackHolder->m_CorProfilerCallback7 = pCorProfilerCallback7;
     }
 
-    CComPtr<ICorProfilerCallback3> pCorProfilerCallback3;
-    hr = pUnkProfilerCallback->QueryInterface(__uuidof(ICorProfilerCallback3), (LPVOID*)&pCorProfilerCallback3);
-    if (SUCCEEDED(hr))
+    // ICorProfilerCallback6
+    if (profilerCallbackHolder->m_CorProfilerCallback7)
+    {
+        profilerCallbackHolder->m_CorProfilerCallback6 = static_cast<CComPtr<ICorProfilerCallback6>>(profilerCallbackHolder->m_CorProfilerCallback7);
```

Comment:
does `static_cast<ICorProfilerCallback6>` give you the same semantics as `static_cast<CComPtr<ICorProfilerCallback6>>`?

Reply:
`static_cast<ICorProfilerCallback*>` seems to work as well

---
In reply to: [491178672](https://github.com/microsoft/CLRInstrumentationEngine/pull/292#discussion_r491178672) [](ancestors = 491178672)